lib/good_job/job.rb

```diff
@@ -193,28 +193,16 @@ def perform(destroy_after: !GoodJob.preserve_job_records, reperform_on_standard_
       raise PreviouslyPerformedError, 'Cannot perform a job that has already been performed' if finished_at
 
       GoodJob::CurrentExecution.reset
-      result = nil
-      rescued_error = nil
-      error = nil
 
       self.performed_at = Time.current
       save! unless destroy_after
 
-      params = serialized_params.merge(
-        "provider_job_id" => id
-      )
-
-      begin
-        ActiveSupport::Notifications.instrument("perform_job.good_job", { good_job: self, process_id: GoodJob::CurrentExecution.process_id, thread_name: GoodJob::CurrentExecution.thread_name }) do
-          result = ActiveJob::Base.execute(params)
-        end
-      rescue StandardError => e
-        rescued_error = e
-      end
+      result, rescued_error = execute
 
       retry_or_discard_error = GoodJob::CurrentExecution.error_on_retry ||
                                GoodJob::CurrentExecution.error_on_discard
 
+      error = nil
       if rescued_error
         error = rescued_error
       elsif result.is_a?(Exception)
```

Comment:
Do you know why/when the result can be an exception? I find it quite surprising.

Reply:
Good question! Apologies if this turns into story time.

Briefly: when using `ActiveJob.rescue_from` (or `retry_on` / `discard_on` which are implemented with `rescue_from`), ActiveJob will return the exception that is caught/handled.

Storytime...

I'm not sure whether the above behavior is intentional or merely a coincidence. 

The behavior is unreliable though. I discovered in production that Sentry's instrumentation interrupts that behavior: https://github.com/getsentry/raven-ruby/blob/a5f101a9c7776168df1c0e2aac13d46b70bbb24e/lib/raven/integrations/rails/active_job.rb#L21-L31

...you'll notice that Sentry will call `rescue_with_handler` but does _not_ return the result. I just checked it off my todo list by submitting a PR (thanks for the inspiration 🎉 ): https://github.com/getsentry/raven-ruby/pull/1027 ...but I have to assume that other gems may also interrupt this behavior too. This is explored in #72.